docs/CADL.md

```diff
@@ -0,0 +1,93 @@
+- InputModel: deserialized object representation of cadl.json in autorest.csharp.
+- OutputModel: variety of objects and object factories used by code writers to generate C# code.
+
+## CADL Pipeline Overview
+
+C# code generation CADL pipeline consists of two major components: CADL compiler that inputs API specification and outputs cadl.json using [CADL C# Emitter](https://github.com/Azure/autorest.csharp/tree/feature/v3/src/CADL.Extension/Emitter.Csharp), and autorest.csharp (a.k.a [generator](https://github.com/Azure/autorest.csharp/tree/feature/v3/src/AutoRest.CSharp)), executed in standalone mode, which inputs cadl.json and outputs C# files that contain  clients and models for interaction with cloud service.
+
+## CADL C# Emitter
+Emitter generates cadl.json from subset of CADL compiler output object graph. Resulting output is a standard JSON file with circular references support. It doesn't have explicit specification. Instead, it is assumed that InputModel types defined in [TS](https://github.com/Azure/autorest.csharp/tree/feature/v3/src/CADL.Extension/Emitter.Csharp/src/type) and in [C#](https://github.com/Azure/autorest.csharp/tree/feature/v3/src/AutoRest.CSharp/Common/Input) match each other. When new features in generator require InputModel types to be modified or extended, there is no backwards compatibility requirement. Hence, releases of autorest.csharp and CADL C# Emitter must be synchronized so that autorest.csharp can deserialize `cadl.json` produced by emitter without errors.
+
+Emitter can apply C#-specific transformations and generation rules to the spec before or in process of cadl.json generation (language-agnostic transformations and rules are expected to be applied to spec before it becomes available to the emitter).
+
+## InputModel vs OutputModel
+
+To represent cadl.json in memory, generator deserializes JSON into InputModel object graph. Then, with respect of configuration settings and customizations extracted from partial types and assembly level attributes (via `SourceInputModel`), generator creates OutputModel object graph. Various entities from OutputModel will be used by code writers to generate C# files. 
+
+InputModel is a slim, unvalidated representation of the cadl.json in memory. OutputModel is a rich, self-consistent set of objects that has all the data code writers need to generate valid C# code.
+
+With the exception of enums, no object should belong to InputModel and OutputModel at the same time.
+
+#### InputModel requirements
+- Minimalistic design. Only nessesary data is available, data duplication should be avoided (e.g. `InputModelType.DiscriminatorPropertyName` makes `InputModelProperty.IsDiscriminator` unnessesary). 
+- All InputModel objects are init-immutable (instance can't be modified after becoming accessible to external code).
+- InputModel is final. All decorator-based and configuration-based transformations and rules must be applied in emitter.
+- Lazy initializations must be avoided.
```

Comment:
why do we highlight this? Is there any problem caused by lazy initialization?

Reply:
Yes. 

Lazy initialization pattern should be used when we don't want to evaluate expression until its value is needed and want to share the result (the other name of the pattern is [call-by-need](https://en.wikipedia.org/wiki/Lazy_evaluation)).

In autorest.csharp, however, it is used for completely different purpose. We know upfront that expression will be called, but we can't call it immediately because it doesn't have all the variables defined. So we declare it lazy, **believing** that we can delay access to the expression value up to the point when all its variables are defined already or can be defined  using similar logic. This is an anti-pattern. In MPG, it has already resulted in code that creates [all types twice](https://github.com/Azure/autorest.csharp/blob/110a403f43e6e6364bba97ced59f9e754d047774/src/AutoRest.CSharp/Mgmt/AutoRest/MgmtOutputLibrary.cs#L748). That is not the worst thing that may happen though. In [Python extension for Visual Studio Code](https://github.com/Microsoft/vscode-python), where lazy initialization has been used for similar purposes (type resolution for IntelliSense and debugging), we got a case of 3 files with 20 lines of code combined that required analyzer more than an hour to process. After several unsuccessful attempts to fix it granularly, I had to rewrite the whole type resolution code. That work took several months.

So I want get rid of lazy initialization misuse while can do it bit by bit, starting by not adding more lazy initializations.